Fix file.line diff formatting.

gvengel opened the pull request:
### What does this PR do?

Fix file.line diff formatting.

Using _splitlines_preserving_trailing_newline() strips off line endings, mangling the unified_diff output. Switching to splitlines with the keepends option resolves this issue, while still preserving trailing newline.

### What issues does this PR fix or reference?

Fixes #41474

### Previous Behavior
Changes diff displays compressed output on a single line.

```
Changes:
         ----------
         diff:
             ---
             +++
             @@ -1,4 +1,5 @@
              egg bacon+sausage spam
```

### New Behavior
Changes diff displays multiline output.

```
Changes:
         ----------
         diff:
             ---
             +++
             @@ -1,3 +1,4 @@
              egg
              bacon
             +sausage
              spam
```

### Tests written?

Yes

### Commits signed with GPG?

Yes

gvengel committed:
Fix file.line diff formatting.

Using _splitlines_preserving_trailing_newline() strips off line
endings, mangling the unified_diff output. Switching to splitlines
with the keepends option resolves this issue, while still
preserving trailing newline.

Fixes #41474

cachedout commented:
I would very much prefer that we could add a test for this before we get this in. @gvengel could you please add an integration test that covers this issue?

gvengel commented:
Sure, I'll find some time to get one added.

gvengel commented:
Should the test go in integration.modules.file.FileModuleTest? Want me to add a docstring referencing the original bug report?

gvengel committed:
Added integration test for issue #41474

gvengel commented:
@cachedout Integration test added. Let me know if it looks good. Thanks.

terminalmage reviewed (approved):
Conditionally I think this is good to merge. The only question I have is whether this may affect how it is processed on Windows. In many cases when we have issues with line endings, differences between Windows and *NIX are to blame. So, this issue may have been caused by over-correcting to fix for Windows, and I want to make sure that whatever we're doing works on both platforms.

gvengel commented:
Ah, good point. I haven't tested this on Windows. I'll get a Windows environment setup, and make sure everything looks good.

rallytime commented on tests/integration/modules/file.py:
```diff
@@ -293,6 +293,16 @@ def test_source_list_for_list_returns_local_file_proto_from_dict(self):
+        '''
+        ret = self.minion_run('file.line', self.myfile, 'Goodbye',
+                              mode='insert', after='Hello')
+        self.assertIn('Hello\n+Goodbye', ret)
```
As mentioned in previous comments, we should be careful with Windows with this change. I haven't tested the code change above, but as this test stands, the test itself will fail on Windows. Something like `os.sep` should be used here. See #43943 for an example.

twangboy replied:
Should be:
```
self.assertIn('Hello' + os.linesep + 'Goodbye', ret)
```
to work on Windows.

gvengel committed:
Fixed test to work on Windows.

twangboy reviewed (changes requested):
Use `os.linesep` instead of hardcoded `\n` for tests in Windows

gvengel commented:
Thanks. I've changed the integration test to use `os.linesep`; however, I've had a hard time actually testing this on Windows. My initial attempts to run the test on Windows resulted in an odd exception. After a few rounds of debugging, I came to the conclusion that file.line is already broken on Windows for reasons unrelated to my change. 

The following test reproduces the issue. I first ran it under the 2016.11 and 2017.7.2 branches in dev mode on Windows 10. Next I ran it using the Salt-Minion-2017.7.2-AMD64-Setup.exe Python 2 and 3 installers from https://docs.saltstack.com/en/latest/topics/installation/windows.html on both Windows 10 and Windows Server 2016. All tests produce the same exception.

```
PS C:\> cat .\test.txt
egg
bacon
spam
PS C:\> salt-call --local file.line test.txt sausage mode=insert after=bacon
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
NameError: global name '_assert_occurrence' is not defined
Traceback (most recent call last):
  File "C:\salt\bin\Scripts\salt-call", line 11, in <module>
    salt_call()
  File "C:\salt\bin\lib\site-packages\salt\scripts.py", line 391, in salt_call
    client.run()
  File "C:\salt\bin\lib\site-packages\salt\cli\call.py", line 58, in run
    caller.run()
  File "C:\salt\bin\lib\site-packages\salt\cli\caller.py", line 134, in run
    ret = self.call()
  File "C:\salt\bin\lib\site-packages\salt\cli\caller.py", line 204, in call
    ret['return'] = func(*args, **kwargs)
  File "C:\salt\bin\lib\site-packages\salt\modules\file.py", line 1778, in line
    _assert_occurrence(body, after, 'after')
NameError: global name '_assert_occurrence' is not defined
Traceback (most recent call last):
  File "C:\salt\bin\Scripts\salt-call", line 11, in <module>
    salt_call()
  File "C:\salt\bin\lib\site-packages\salt\scripts.py", line 391, in salt_call
    client.run()
  File "C:\salt\bin\lib\site-packages\salt\cli\call.py", line 58, in run
    caller.run()
  File "C:\salt\bin\lib\site-packages\salt\cli\caller.py", line 134, in run
    ret = self.call()
  File "C:\salt\bin\lib\site-packages\salt\cli\caller.py", line 204, in call
    ret['return'] = func(*args, **kwargs)
  File "C:\salt\bin\lib\site-packages\salt\modules\file.py", line 1778, in line
    _assert_occurrence(body, after, 'after')
NameError: global name '_assert_occurrence' is not defined
PS C:\>
```

The same test works fine on Linux.

```
root@ubuntu-xenial:~# cat test.txt 
egg
bacon
spam
root@ubuntu-xenial:~# salt-call --local file.line test.txt sausage mode=insert after=bacon
local:
    --- 
    +++ 
    @@ -1,3 +1,4 @@
     egg
     bacon
    +sausage
     spam
root@ubuntu-xenial:~# 
```

I'm quite mystified by this issue. The _assert_occurrence() function is clearly defined above line() in file.py on line 1568 (2017.7.2). It is accessible in the global scope on Linux, so why is Windows any different? 🙀

twangboy commented:
@gvengel 

Windows has it's own module (`win_file.py`) that "namespaces" functions from the `file.py` module. So, that function needs to be "namespaced" in the `win_file.py` module. You're right, it probably wasn't working before.

To "namespace" the function in `win_file.py` you'll need to do the following:

1. Add the `_assert_occurrence` function to the import from `file.py`. The line starts with:
   `from salt.modules.file import`

2. Create a global variable for the function in the `__virtual__` function by adding as a variable it to one of the `global` lines

3. "Namespace" the function by adding the line:
   `_assert_occurrence = _namespaced_function(_assert_occurrence, globals())`
   to the list of namespaced functions in the `__virtual__` function.

That should do it.

gvengel commented:
@twangboy Ah, I knew it must be something like that. Should I roll that change into mine, or do you want a separate PR and/or bug report?

twangboy commented:
@cachedout ^^^^

gvengel commented:
I've stashed the change to namespace the required functions, pending feedback.

Unfortunately, it looks like file.line is still busted on Windows. This is unrelated to my fix for the changes diff. The diff displays correctly, and my integration test passes, but the file on disk gets an extra carriage return inserted into every line.

I'll find some time to dig into and fix that too. 😄

gvengel commented:
Okay, the problem is with mixing salt.utils.fopen() to read the file and then salt.utils.atomicfile.atomic_open() to write it back out. You can see fopen() has a special case to switch the mode to binary, but then the subsequent write with atomic_open() is in text mode.

Excerpt from fopen():
```
    # ensure 'binary' mode is always used on Windows in Python 2
    if ((six.PY2 and is_windows() and 'binary' not in kwargs) or
            kwargs.pop('binary', False)):
```

I've stashed a fix for this too. Just let me know if it's okay to include the two additional bug fixes in this PR, or you want a separate "fix file.line on windows" PR. @cachedout

rallytime commented:
@gvengel Fixing the windows issue is fine to include with this fix. Thank you so much for tackling that.

gvengel commented:
Great, that makes things easier. 

I think we'll want an integration test to validate the file content, in addition to the diff formatting. I'll get that added and update the PR this weekend.

rallytime commented:
@gvengel That would be wonderful if you could add an additional integration test for that.

cachedout commented:
Sorry for the delay here. This all sounds just fine and as soon as everything is rolled into this, I think we can likely go ahead and just get it in.

gvengel committed:
Namespace missing functions for file.line on Windows.

gvengel committed:
Fix FileModuleTest setUp and tearDown to work on Windows.

gvengel committed:
Fixed issue with file.line on Windows running Python 2.

The target file was being read as binary, but written as text, causing file.line to append an extra linesep on every line.

Added integration test to ensure file.line produces the expected content.

gvengel commented:
The integration tests have been added, along with fixes to get them running on Windows.

Part of the problem was that `tests.integration.modules.file` was trying to import Posix only modules on Windows, which caused runtests.py to silently fail to load all the tests in `FileModuleTest`. Possibly a bug in runtests.py causing it to catch and ignore ImportError?

Now that all the`FileModuleTest` tests run on Windows, there are still a number which are failing. I can take a look at those too, but I focused on getting `file.line` working so far.

Commit:
Add newline for lint.

rallytime reviewed (approved):
Thanks @gvengel!
